elasticapm/utils/disttracing.py

```diff
@@ -79,7 +83,42 @@ def from_string(cls, traceparent_string):
         except ValueError:
             logger.debug("Invalid trace-options field, value %s", trace_flags)
             return
-        return TraceParent(version, trace_id, span_id, tracing_options)
+        return TraceParent(version, trace_id, span_id, tracing_options, tracestate_string, is_legacy)
+
+    @classmethod
+    def from_headers(
+        cls,
+        headers,
+        header_name=constants.TRACEPARENT_HEADER_NAME,
+        legacy_header_name=constants.TRACEPARENT_LEGACY_HEADER_NAME,
+        tracestate_header_name=constants.TRACESTATE_HEADER_NAME,
+    ):
+        tracestate = cls.merge_duplicate_headers(headers, tracestate_header_name)
+        if header_name in headers:
+            return TraceParent.from_string(headers[header_name], tracestate, is_legacy=False)
+        elif legacy_header_name in headers:
+            return TraceParent.from_string(headers[legacy_header_name], tracestate, is_legacy=False)
+        else:
+            return None
+
+    @classmethod
+    def merge_duplicate_headers(cls, headers, key):
+        """
+        HTTP allows multiple values for the same header name. Most WSGI implementations
+        merge these values using a comma as separator (this has been confirmed for wsgiref,
+        werkzeug, gunicorn and uwsgi). Other implementations may use containers like
+        multidict to store headers and have APIs to iterate over all values for a given key.
+
+        This method is provided as a hook for framework integrations to provide their own
+        TraceParent implementation. The implementation should return a single string. Multiple
+        values for the same key should be merged using a comma as separator.
+
+        :param headers: a dict-like header object
+        :param key: header name
+        :return: a single string value
+        """
+        # this works for all known WSGI implementations
+        return headers.get(key)
```

Comment:
Good work being so thorough with your investigation here. And for providing something to override if Murphy's Law gives us a WSGI implementation with different behavior.